### PR TITLE
feat: delete user attributes

### PR DESCRIPTION
**Pull request by hideokamoto**
## _Issue #, if available:_
Ref: Expose promisifed helper to delete attributes on Auth #3119
Currently, there is no method to delete the user attributes in the Auth class.

## _Description of changes:_
I've added the method into the class.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -1102,6 +1102,35 @@ export class AuthClass {
...
 
+	/**
+	 * Delete an authenticated users' attributes
+	 * @param {CognitoUser} - The currently logged in user object
+	 * @return {Promise}
+	 **/
+	public deleteUserAttributes(
+		user: CognitoUser | any,
+		attributeNames: string[],
+		clientMetadata: ClientMetaData = this._config.clientMetadata
```
Comment (hkjpotato) [inline packages/auth/src/Auth.ts:1113]: can you remove the `clientMetadata`, as it is not accepted by the [`deleteAttributes` function](https://github.com/aws-amplify/amplify-js/blob/main/packages/amazon-cognito-identity-js/src/CognitoUser.js#L1327). Behind the scene, [Coginto.DeleteUserAttributes](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_DeleteUserAttributes.html#API_DeleteUserAttributes_RequestSyntax) does not respect this input.

**Review by hkjpotato — COMMENTED** [review]
Thank you for submitting the PR!

File: packages/auth/src/Auth.ts (hunk 1 above)
Comment (hideokamoto) [reply packages/auth/src/Auth.ts]: Thanks for the feedback!
I've updated the code to delete it :)

**Comment by hkjpotato** [comment]
since this PR https://github.com/aws-amplify/amplify-js/pull/8578 has been merged, I am good with the new deleteUserAttributes API on Auth class.

@hideokamoto can you add unit tests for the new `deleteUserAttributes` method, like the sibling `updateUserAttributes` method. Basically,
- You need to mock the deleteAttributes call on CognitoUser, like [this](https://github.com/aws-amplify/amplify-js/blob/main/packages/auth/__tests__/auth-unit-test.ts#L204).
- Then add unit tests for the Auth `deleteUserAttributes` method, like [this](https://github.com/aws-amplify/amplify-js/blob/main/packages/auth/__tests__/auth-unit-test.ts#L2623).

**Comment by hideokamoto** [comment]
The test command can not run.

```bash
% yarn run test --scope @aws-amplify/auth
yarn run v1.22.10
$ lerna run test --stream --scope @aws-amplify/auth
lerna notice cli v3.22.1
lerna info versioning independent
lerna notice filter including "@aws-amplify/auth"
lerna info filter [ '@aws-amplify/auth' ]
lerna info Executing command in 1 package: "yarn run test"
@aws-amplify/auth: $ yarn lint --fix && jest -w 1 --coverage
@aws-amplify/auth: $ tslint '{__tests__,src}/**/*.ts' --fix
@aws-amplify/auth: ts-jest[config] (WARN) TypeScript diagnostics (customize using `[jest-config].globals.ts-jest.diagnostics` option):
@aws-amplify/auth: message TS151001: If you have issues related to imports, you should consider setting `esModuleInterop` to `true` in your TypeScript configuration file (usually `tsconfig.json`). See https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/#easier-ecmascript-module-interoperability for more information.
@aws-amplify/auth: <--- Last few GCs --->
@aws-amplify/auth: [17977:0x108008000]   102392 ms: Mark-sweep (reduce) 4093.0 (4102.3) -> 4092.3 (4103.3) MB, 4721.1 / 0.0 ms  (average mu = 0.120, current mu = 0.003) allocation failure scavenge might not succeed
@aws-amplify/auth: [17977:0x108008000]   107467 ms: Mark-sweep (reduce) 4093.3 (4105.3) -> 4092.8 (4105.8) MB, 5066.0 / 0.0 ms  (average mu = 0.060, current mu = 0.002) allocation failure scavenge might not succeed
@aws-amplify/auth: <--- JS stacktrace --->
@aws-amplify/auth: FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
@aws-amplify/auth:  1: 0x10130d5e5 node::Abort() (.cold.1) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  2: 0x1000b2289 node::Abort() [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  3: 0x1000b23ef node::OnFatalError(char const*, char const*) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  4: 0x1001f68c7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  5: 0x1001f6863 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  6: 0x1003a47e5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  7: 0x1003a628a v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  8: 0x1003a19b5 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth:  9: 0x10039f2e0 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 10: 0x1003ad9ea v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 11: 0x1003ada71 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 12: 0x100374fc7 v8::internal::FactoryBase<v8::internal::Factory>::NewFixedArrayWithFiller(v8::internal::Handle<v8::internal::Map>, int, v8::internal::Handle<v8::internal::Oddball>, v8::internal::AllocationType) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 13: 0x10037c643 v8::internal::Factory::NewFrameArray(int) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 14: 0x1003416a3 v8::internal::(anonymous namespace)::CaptureStackTrace(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::(anonymous namespace)::CaptureStackTraceOptions) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 15: 0x100341e61 v8::internal::Isolate::CaptureAndSetSimpleStackTrace(v8::internal::Handle<v8::internal::JSReceiver>, v8::internal::FrameSkipMode, v8::internal::Handle<v8::internal::Object>) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 16: 0x10028a610 v8::internal::Builtin_Impl_ErrorCaptureStackTrace(v8::internal::BuiltinArguments, v8::internal::Isolate*) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 17: 0x100a82099 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_BuiltinExit [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: 18: 0x100a1ae42 Builtins_InterpreterEntryTrampoline [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
@aws-amplify/auth: /bin/sh: line 1: 17977 Abort trap: 6           jest -w 1 --coverage
@aws-amplify/auth: error Command failed with exit code 134.
@aws-amplify/auth: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
lerna ERR! yarn run test exited 134 in '@aws-amplify/auth'
```

**Comment by hideokamoto** [comment]
@hkjpotato

I can't run the test code in my local.
No way to resolve the HEAP Error.

How can we resolve it?

```bash
node --trace-gc --max-old-space-size=20240   ./node_modules/.bin/lerna run test --scope @aws-amplify/auth

lerna ERR! yarn run test stderr:
ts-jest[config] (WARN) TypeScript diagnostics (customize using `[jest-config].globals.ts-jest.diagnostics` option):
message TS151001: If you have issues related to imports, you should consider setting `esModuleInterop` to `true` in your TypeScript configuration file (usually `tsconfig.json`). See https://blogs.msdn.microsoft.com/typescript/2018/01/31/announcing-typescript-2-7/#easier-ecmascript-module-interoperability for more information.

<--- Last few GCs --->

[19730:0x108008000]    99319 ms: Mark-sweep (reduce) 4095.0 (4106.7) -> 4094.9 (4107.7) MB, 4106.4 / 0.0 ms  (+ 87.8 ms in 16 steps since start of marking, biggest step 15.1 ms, walltime since start of marking 4205 ms) (average mu = 0.071, current mu = 0.[19730:0x108008000]   103623 ms: Mark-sweep (reduce) 4095.9 (4104.7) -> 4095.5 (4106.5) MB, 4268.8 / 0.0 ms  (+ 24.8 ms in 16 steps since start of marking, biggest step 13.6 ms, walltime since start of marking 4305 ms) (average mu = 0.036, current mu = 0.

<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x10130d5e5 node::Abort() (.cold.1) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 2: 0x1000b2289 node::Abort() [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 3: 0x1000b23ef node::OnFatalError(char const*, char const*) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 4: 0x1001f68c7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 5: 0x1001f6863 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 6: 0x1003a47e5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 7: 0x1003a628a v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 8: 0x1003a19b5 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
 9: 0x10039f2e0 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
10: 0x1003ad9ea v8::internal::Heap::AllocateRawWithLightRetrySlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
11: 0x1003ada71 v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
12: 0x10037b862 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
13: 0x1006fbb38 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
14: 0x100a81fb9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/Users/okamotohidetaka/.anyenv/envs/nodenv/versions/14.17.1/bin/node]
/bin/sh: line 1: 19730 Abort trap: 6           jest -w 1 --coverage
error Command failed with exit code 134.

lerna ERR! yarn run test exited 134 in '@aws-amplify/auth'
```

**Comment by hkjpotato** [comment]
@hideokamoto thank you for adding the test, nice work! Your test actually passes when run by the ci/circleci, I also checkout your change locally and verify it passes.

I haven't seen the heap error you show before, have you follow https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions to run `yarn run test --scope @aws-amplify/auth`?

I will ping my teammate to see if we can merge this PR now, I might need add some doc change as well.

**Review by hkjpotato — APPROVED** [review]

**Review by elorzafe — APPROVED** [review]
Thank you @hideokamoto 🌮 🎉 🥇

**Comment by hkjpotato** [comment]
doc update PR: https://github.com/aws-amplify/docs/pull/3413

**Pull request merged**